src/System.IO.Pipelines/src/System/IO/Pipelines/Pipe.cs

```diff
@@ -0,0 +1,758 @@
+
+        private PipeCompletion _writerCompletion;
+        private PipeCompletion _readerCompletion;
+
+        // The read head which is the extent of the IPipelineReader's consumed bytes
+        private BufferSegment _readHead;
+        private int _readHeadIndex;
+
+        // The commit head which is the extent of the bytes available to the IPipelineReader to consume
+        private BufferSegment _commitHead;
+        private int _commitHeadIndex;
+
+        // The write head which is the extent of the IPipelineWriter's written bytes
+        private BufferSegment _writingHead;
+
+        private PipeReaderState _readingState;
+
+        private bool _disposed;
+
+        internal long Length => _length;
+
+        /// <summary>
+        /// Initializes the <see cref="Pipe"/> using <see cref="PipeOptions.Default"/> as options.
+        /// </summary>
+        public Pipe() : this(PipeOptions.Default)
+        {
+        }
+
+        /// <summary>
+        /// Initializes the <see cref="Pipe"/> with the specified <see cref="PipeOptions"/>.
+        /// </summary>
+        public Pipe(PipeOptions options)
+        {
+            if (options == null)
+            {
+                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.options);
+            }
+
+            _bufferSegmentPool = new BufferSegment[SegmentPoolSize];
+
+            _readingState = default;
+            _readerCompletion = default;
+            _writerCompletion = default;
+
+            _pool = options.Pool;
+            _minimumSegmentSize = options.MinimumSegmentSize;
+            _pauseWriterThreshold = options.PauseWriterThreshold;
+            _resumeWriterThreshold = options.ResumeWriterThreshold;
+            _readerScheduler = options.ReaderScheduler ?? PipeScheduler.Inline;
+            _writerScheduler = options.WriterScheduler ?? PipeScheduler.Inline;
+            _readerAwaitable = new PipeAwaitable(completed: false);
+            _writerAwaitable = new PipeAwaitable(completed: true);
+            _reader = new DefaultPipeReader(this);
+            _writer = new DefaultPipeWriter(this);
+        }
+
+        private void ResetState()
+        {
+            _readerCompletion.Reset();
+            _writerCompletion.Reset();
+            _commitHeadIndex = 0;
+            _currentWriteLength = 0;
+            _length = 0;
+        }
+
+        /// <summary>
+        /// Allocates memory from the pipeline to write into.
+        /// </summary>
+        /// <param name="minimumSize">The minimum size buffer to allocate</param>
+        internal Memory<byte> GetMemory(int minimumSize)
+        {
+            if (_writerCompletion.IsCompleted)
+            {
+                ThrowHelper.ThrowInvalidOperationException_NoWritingAllowed();
+            }
+
+            if (minimumSize < 0)
+            {
+                ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument.minimumSize);
+            }
+
+            lock (_sync)
```

Comment:
All of these `lock`s expressions will now throw compiler error `CS0656` saying:
```
System\IO\Pipelines\Pipe.cs(130,13): error CS0656: Missing compiler required member 'System.Threading.Monitor.Exit' [F:\git\corefx\src\System.IO.Pipelines\src\System.IO.Pipelines.csproj]
System\IO\Pipelines\Pipe.cs(130,13): error CS0656: Missing compiler required member 'System.Threading.Monitor.Enter' [F:\git\corefx\src\System.IO.Pipelines\src\System.IO.Pipelines.csproj]
```

Reply:
Fixed by adding System.Threading reference.